text/metrics/0238-pipeline-monitoring.md

```diff
@@ -0,0 +1,364 @@
+# OpenTelemetry Telemetry Pipeline metrics
+
+Propose a uniform standard for telemetry pipeline metrics generated by
+OpenTelemetry SDKs and Collectors.
+
+WIP: this is a work-in-progress draft, derived from OTEP 238 which
+was closed pending further development.
+
+Tl;DR: this proposes TWO metric instrument semantics, three overall
+considering SDKs and collectors:
+
+- `otelcol_consumed_items`: Received and inserted data items (Collector)
+- `otelcol_produced_items`: Exported, dropped, and discarded items (Collector)
+- `otelsdk_produced_items`: Exported, dropped, and discarded items (SDK)
+
+## Motivation
+
+OpenTelemetry desires to standardize conventions for the metrics
+emitted by SDKs about success and failure of telemetry reporting. At
+the same time, the OpenTelemetry Collector has existing conventions
+which are expected to connect with metrics emitted by SDKs and have
+similar definitions.
+
+## Explanation
+
+We use the term "pipeline" to describe an arrangement of system
+components which produce, consume, and process telemetry on its way
+from the point of origin to the endpoint(s) in its journey.  Pipeline
+components included in this specification are:
+
+- OpenTelemetry SDKs: As telemetry producers, these components are the
+  start of a pipeline.
+- OpenTelemetry Collectors: The OpenTelemetry collector contains an
+  arrangement of components which act as both consumers and producers.
+  
+The terms "following"/"follower" and "preceding"/"preceder" refer to
+the relative components after or before another component in a
+pipeline.  The preceding component ("preceder") produces data that is
+consumed by the following component ("follower").
+
+An arrangement of pipeline components acting as a single unit, such as
+implemented by the OpenTelemetry Collector, is called a segment.  Each
+segment consists of a receiver, zero or more processors, and an
+exporter.  The terms "following" and "preceding" apply to pipeline
+segments with the same meaning as for components.  For example, a
+agent pipeline segment forwards to a gateway pipeline.
+
+### Detailed design
+
+#### Producer and Consumer instruments
+
+We choose to specify two metric instruments for use counting outcomes,
+one instrument to account for producer outcomes and one to account for
+consumer outcomes.  In an ideal pipeline, a conservation rule exists
+between what goes in (i.e., is consumed) and what goes out (i.e., is
+produced).  The use of producer and consumer metric instruments is
+designed to enable this form of consistency check.  When the pipeline
```

Comment:
From an accounting perspective, I see why we would want to group received + inserted items (so that this total matches exported + dropped + discarded). But the language here is difficult to reconcile with the external vs internal nature of the operations.

Taking a step back, I agree with the categories you've identified (received, exported, inserted, discarded, dropped), but there are several ways to organize them. This proposal organizes the categories in terms of incremental (received, inserted) vs decremental (discarded, dropped, exported) because it gives us the desirable property that the two instruments should be equal. However, I wonder if these same categories can be modeled in a different way while still giving us the ability to check consistency.

Would it be enough that all categories should sum to 0 by subtracting the decremental operations from the incremental ones? Organized according to real data flow, it would be `received - discarded + inserted - (dropped + exported) = 0`. I think that by separating the incremental from the decremental, it allows this to work for backends, but alternately, could we require that the decremental categories are reported as negative numbers within the same instrument? To me this seems more intuitive but I'm not sure all backends can handle this.

Reply:
I'm happy with the equation `received - discarded + inserted - (dropped + exported) = 0`.

I don't think I see a difference between `received` and `inserted`. If the telemetry has the component name, it'll be clear whether it was a processor or a receiver, and could be just a semantic question. If we added another attribute to identify the _kind_ of component, or required it to be included in the `otel.component` attribute, is that enough to distinguish received and inserted?

My thinking, in creating a `discarded` and `dropped` designation specifically was to have enough decomposition in the data that you could perform the equation as you wrote it, meaning to count received (receivers), subtract discarded, add received (processors), subtract dropped, leaving exported, which is the thing you'll compare with the next segment, potentially.

Reply:
Continuing --

Your suggestion about negative-values, as opposed to the positive-only expression I've used, brings to mind several related topics. I think this is the "best" way to do it from a metrics data model perspective, but I want to point out other ways we can map these metric events.

Consider each item of telemetry that enters the pipeline, has with it an associated trace context. There is:

a. The UpDownCounter formulation -- for every item arriving, add 1. for every item departing, subtract 1. this can tell us the number of items for attribute sets that are symmetric.  If we add one for every item that is input/consumed, then subtract one for every item that is output/produced, the resulting tally is a number of in-flight items, but this mapping has to ignore the outcome/success labels for the +1/-1 to balance out. 
b. The Span formulation -- when the receiver starts a new request (or the processor inserts some new data), there is an effective span start event (or a log about the arrival of some telemetry) for some items of telemetry. When the outcome is known for those points (having called the follower), there is a span finish event which can be annotated w/ the subtotal for each outcome/success matching the number of items consumed.
c. The LogRecord formulation -- (same as span formula, but one log record per event, vs span start/end events).

I'm afraid to keep adding text to the document, but I would go further with the above suggestions. If we are using metrics to monitor the health of all the SDKs, then we will be missing a signal when the metrics SDK itself is failing. I [want the metrics SDK to have a span](https://github.com/lightstep/otel-launcher-go/blob/12591d57c67148abaf7e74b0bbeacf5c18d1f023/lightstep/sdk/metric/exporters/otlp/otelcol/client.go#L216) encapsulating each export operation.

Reply:
> I don't think I see a difference between received and inserted. If the telemetry has the component name, it'll be clear whether it was a processor or a receiver, and could be just a semantic question. If we added another attribute to identify the kind of component, or required it to be included in the otel.component attribute, is that enough to distinguish received and inserted?

Looks like I missed an important part of the design: `processors are responsible for counting items only when the number changes while passing through a processor`

I was thinking was that we should report "received" and "exported" for processors in order to account for situations where data streams are merged. For example, a collector pipeline with two receivers will combine streams into the first processor, so from that processor's perspective it seems important to report the total "received". Likewise, a similar problems could arise from receivers or exporters used in multiple pipelines. 

To use a concrete example:
```yaml
pipelines:
  logs/1:
    receivers: [R1, R2]
    processors: [P1]
    exporters: [E1, E2]
  logs/2:
    receivers: [R1]
    processors: [P2]
    exporters: [E1]
```

| component | received | discarded | inserted | dropped | exported |
| --- | --- | --- | --- | --- | --- |
| R1 | 10 | - | - | - | - |
| R2 | 20 | - | - | - | - |
| P1 | 30 | 25 | 0 | - | 5 |
| P2 | 10 | 10 | 2 | - | 2 |
| E1 | - | - | - | 0 | 7 |
| E2 | - | - | - | 0 | 5 |

In this example, it seems much easier to understand what's going on with `P1` when it reports receiving 30.

Reply:
My earlier design for this proposal included what you're suggesting -- the idea that every processor in the pipeline will independently report complete totals. I think is excessive, there is a lot of redundancy, but the problem can be framed this way. In fact, the current design can be applied the way you describe by a simple redefinition rule -- if you consider a pipeline segment to be an individual receiver, an individual processor, or an individual exporter, you'll get the metrics you're expecting. I think this might even be appropriate in complex pipelines.

The defect I'm aware of, when each processor counts independent totals, is that it becomes easy to aggregate adjacent pipeline segments together, which results in overcounting from a pipeline perspective. This is not a unique problem to processor metrics -- the problem arises when a metric query aggregates more than one collector belonging to the same pipeline, or more than one exporter, or more than one processor. My goal is to make it easy to write queries that encompass 

In my current proposal, if you aggregate the total for `otelcol_consumed_items` grouping by all attributes to a single total, the result will be the number of collector pipeline segments times the number of items. If you restrict your query to one segment (meaning one pipeline and one collector), then the aggregate equals the number of items. This property holds because each segment has one exporter and one receiver. 

Since there are multiple processors in a pipeline segment, if each processor counts a total, then the aggregate for that segment will equal the number of processors times the number of items, which is not a useful measure to compare against adjacent pipeline segments. When each processor reports a total, you have to aggregate down to an individual processor to understand its behavior. But then, the logic to check whether the receiver and exporter are consistent, given processor behavior, becomes complicated at best--the aggregation would have to filter the `dropped` and `discarded` categories from the processor metrics, and then we'd be able to recover the pipeline equations in this proposal.

This is why I ended up proposing that processors count changes in item count, because the changes in item count aggregate correctly despite multiple processors.

Reply:
> an individual processor, or an individual exporter, you'll get the metrics you're expecting. I think this might even be appropriate in complex pipelines.

Just to be clear, I'm serious about this alternative because it's the same design, just applied at a finer level of detail. I think it might even be reasonable to let this be a runtime configuration, saying whether to export component-level metrics or segment-level metrics. I prefer segment-level metrics for all the reasons written in this document, but component-level metrics also work and can use the same metric instrument names. 

As an rough example, the change of behavior would be:
- Let every component count both consumed and produced (or input and output) items
- Replace `otel.pipeline` attribute with the pipeline name and the component name

I suggest this because you asked "what's going on with `P1`". There is a lot of redundancy, and while it may make it easier to understand the individual component, it seems to make it more difficult to understand the pipeline as a whole.

Reply:
Thanks for explaining further. The tradeoffs are tough here but if we're defining a segment as having only one receiver and one exporter, it excludes a large percentage (maybe substantial majority?) of collector configurations. Even in a simple pipeline like below change counts for `P1` have little meaning.
```yaml
    receivers: [R1, R2]
    processors: [P1]
    exporters: [E1]
```

Reply:
Question about the example, specifically.
Why are there two paths between R1 and E1? This fact will make it difficult to monitor the pipeline, because it appears to double the input on purpose. The pipeline equations will show this happening, but it will be up to interpretation to say whether it's on purpose or not.

The way I would monitor the setup in your example is to compute all the paths for which I expect the conservation rule to hold. They are:

(R1 + R2) -> P1 -> E1
(R1 + R2) -> P1 -> E2
R1 -> P2 -> E1

Since two paths lead to E1, the pipeline equations have to be combined. For E1, the equation will include a factor of 2 for R1.

2*Received(R1) + Received(R2) = Dropped(P1) + Dropped(P2) + Exported(E1)

This kind of calculation can be automated and derived from the metrics I'm proposing, if you have the graph. I mean, if you want to know that P1 received 30 items of telemetry, just add R1 and R2's consumed item totals, that should be easy.

Reply:
> we're defining a segment as having only one receiver and one exporter

This is an interesting statement -- I've definitely not been clear on this topic. I didn't mean to say one receiver and one exporter. I meant all receivers and one exporter, because that's where the conservation principle holds. The sum of all receivers reaches every exporter, and that is a pipeline segment, so your second example,

```
    receivers: [R1, R2]
    processors: [P1]
    exporters: [E1]
```

is exactly the kind of simple pipeline segment that will be easy to monitor, and it will be easy to monitor even if it has a bunch of processors too.

Reply:
> Why are there two paths between R1 and E1?

I agree it's likely not useful. It's a contrived example but I wanted to include the full set of possible stream merges and fan outs:
- single pipeline concerns
  - merge before first processor
  - fan out after last processors
- inter-pipeline concerns
  - fan out after receiver shared by multiple pipelines
  - merge before exporter shared by multiple pipelines

> (R1 + R2) -> P1 -> E1
> (R1 + R2) -> P1 -> E2

I think is perhaps where I'm getting tripped up. Could we define a segment as being able to have more than one receiver? This still aggregates correctly. I see why we cannot include multiple exporters, because data is fanned out _within the segment_, but the fanout that occurs when a receiver is shared between pipelines does not affect the counts for an individual pipeline.

Reply:
> I didn't mean to say one receiver and one exporter. I meant all receivers and one exporter, because that's where the conservation principle holds.

I commented before seeing this but I see we arrived at the same conclusion. 👍